helix-term/src/application.rs

```diff
@@ -200,40 +213,65 @@ impl Application {
             signals,
             jobs: Jobs::new(),
             lsp_progress: LspProgressMap::new(),
+            last_render: Instant::now(),
         };
 
         Ok(app)
     }
 
     fn render(&mut self) {
-        let mut cx = crate::compositor::Context {
-            editor: &mut self.editor,
-            jobs: &mut self.jobs,
-            scroll: None,
-        };
+        #[cfg(feature = "integration")]
+        return;
 
-        self.compositor.render(&mut cx);
-    }
+        #[allow(unreachable_code)]
+        {
+            let compositor = &mut self.compositor;
 
-    pub async fn event_loop(&mut self) {
-        let mut reader = EventStream::new();
-        let mut last_render = Instant::now();
-        let deadline = Duration::from_secs(1) / 60;
+            let mut cx = crate::compositor::Context {
+                editor: &mut self.editor,
+                jobs: &mut self.jobs,
+                scroll: None,
+            };
+
+            compositor.render(&mut cx);
+        }
+    }
 
+    pub async fn event_loop<S>(&mut self, input_stream: &mut S)
+    where
+        S: Stream<Item = crossterm::Result<crossterm::event::Event>> + Unpin,
+    {
         self.render();
+        self.last_render = Instant::now();
 
         loop {
             if self.editor.should_close() {
                 break;
             }
 
+            self.event_loop_until_idle(input_stream).await;
+        }
+    }
+
+    pub async fn event_loop_until_idle<S>(&mut self, input_stream: &mut S) -> bool
+    where
+        S: Stream<Item = crossterm::Result<crossterm::event::Event>> + Unpin,
+    {
+        #[cfg(feature = "integration")]
+        let mut idle_handled = false;
+
+        loop {
+            if self.editor.should_close() {
+                return false;
+            }
```

Comment:
Why do we need to run `should_close()` twice?

Reply:
Because the user could still close all the views while it's in this loop. In this function specifically, without this check, it would have to wait the whole idle timeout before it would exit. It seems prudent to check if it should close right away. I don't think it's an expensive check.

Reply:
Ah, I noticed both `should_close` perform different operation, I wonder why not we do
```rust
if !self.event_loop_until_idle(input_stream).await {
    break;
}
```
in the outer loop? Since the inner loop will return a boolean.

Reply:
Oh I see what you mean. Yeah, that seems logically equivalent. I'll make this change.